File: lintcode/107-Word-Break/solution.java
public class Solution {
    /**
     * @param s: A string s
     * @param dict: A dictionary of words dict
     */
    public boolean wordBreak(String s, Set<String> dict) {
        if (s == null || s.length() == 0) {
            return true;
        }

        // dp[i]: if 0 -> i can be divided
        // dp[i] = dp[k] && dict.contains(s.substring(k + 1, i + 1))
        boolean[] dp = new boolean[s.length() + 1];
        dp[0] = true;
        int maxLength = getMaxLength(dict);

        for (int i = 1; i < dp.length; i++) {
            dp[i] = false;
            for (int j = i - 1; j >= 0 && i - j <= maxLength; j--) {
                if (dp[j] == false) {
                    continue;
                }
                if (dict.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[dp.length - 1];
    }

    private int getMaxLength(Set<String> dict) {
        int maxLength = 0;
        for (String s : dict) {
            maxLength = Math.max(maxLength, s.length());
        }

        return maxLength;
    }
}

File: lintcode/107-Word-Break/solution.cpp
class Solution {
public:
    /**
     * @param s: A string s
     * @param dict: A dictionary of words dict
     */
    bool wordBreak(string s, unordered_set<string> &dict) {
        // dp(i): the position we cut can be word
        // dp(i) = true when dp(j) = true && dict[s[i - j]] = true
        vector<bool> dp(s.size() + 1, false);
        dp[0] = true; // empty string
        
        int max_length = 0;
        for (auto it : dict) {
            if (it.size() > max_length) {
                max_length = it.size();
            }
        }

        // fill dp
        for (int i = 1; i < dp.size(); i++) {
            // from right to left, we don't need to search more than longest word
            for (int j = i - 1; j >= 0 && j >= i - max_length; j--) {
                if (dict.find(s.substr(j, i - j)) != dict.end() && dp[j]) {
                    dp[i] = true;
                    break;
                }
            }
        }
        
        return dp[dp.size() - 1];
    }
};
